agent/grpc-middleware/rate.go

```diff
@@ -0,0 +1,65 @@
+package middleware
+
+import (
+	"context"
+	"errors"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+	"google.golang.org/grpc/tap"
+
+	recovery "github.com/grpc-ecosystem/go-grpc-middleware/recovery"
+
+	"github.com/hashicorp/consul/agent/consul/rate"
+)
+
+// ServerRateLimiterMiddleware implements a ServerInHandle function to perform
+// RPC rate limiting at the cheapest possible point (before the full request has
+// been decoded).
+func ServerRateLimiterMiddleware(limiter RateLimiter, panicHandler recovery.RecoveryHandlerFunc) tap.ServerInHandle {
+	return func(ctx context.Context, info *tap.Info) (_ context.Context, retErr error) {
+		// This function is called before unary and stream RPC interceptors, so we
+		// must handle our own panics here.
+		defer func() {
+			if r := recover(); r != nil {
+				retErr = panicHandler(r)
+			}
+		}()
+
+		peer, ok := peer.FromContext(ctx)
+		if !ok {
+			// This should never happen!
+			return ctx, status.Error(codes.Internal, "gRPC rate limit middleware unable to read peer")
+		}
```

Comment:
What about the case where we do internal gRPC (http to gRPC inside the same server), does it have the correct peer info?

Reply:
From what I can tell, there's no equivalent of the net/rpc in-memory codec in our gRPC stack (we dial a real TCP connection) so the peer here will be whichever server agent is handling the HTTP request.

I guess that's probably ok in this phase?

Reply:
If we do need this, can we avoid blocking the PR on it?  So can we add either a Phase 1 or Phase 2 follow up task if this is needed?

Reply:
+1 to not blocking the PR over this in all cases!